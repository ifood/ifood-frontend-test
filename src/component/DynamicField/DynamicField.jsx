import React from 'react';
import PropTypes from 'prop-types';
import { Select } from 'semantic-ui-react';
import DatePicker from 'react-datepicker';
import { useIntl } from 'react-intl';
import Field from '../Field';
import {
  getComponentType,
  SUPORTED_TYPE,
  selectValuesMapper,
} from './FieldMetadata';
import { getLocale } from '../../core/l10n/L10n';

function DynamicField({ metadata, value = '', onChange }) {
  const intl = useIntl();

  if (metadata == null) {
    return null;
  }

  function handleChange(data) {
    if (onChange != null) {
      onChange(data);
    }
  }

  const handleChangeSelect = (field) => (e, { value: newValue }) => {
    handleChange({ field, value: newValue });
  };

  const handleChangeInput = (field) => (e) => {
    handleChange({
      field,
      value: e?.target?.value,
    });
  };

  const handleChangeDate = (field) => (newValue) => {
    handleChange({ field, value: newValue });
  };

  function getComponentByMetadata(fieldMetadata) {
    const { values, name, id, validation } = fieldMetadata;
    const type = getComponentType(fieldMetadata);

    switch (type) {
      case SUPORTED_TYPE.SELECT:
        return [
          Select,
          {
            fluid: true,
            options: selectValuesMapper({ id, values }),
            placeholder: name,
            value,
            onChange: handleChangeSelect(id),
          },
        ];
      case SUPORTED_TYPE.NUMBER:
        return [
          'input',
          {
            id,
            placeholder: name,
            type: 'number',
            min: validation?.min,
            max: validation?.max,
            value,
            onChange: handleChangeInput(id),
          },
        ];
      case SUPORTED_TYPE.DATE_TIME:
        return [
          DatePicker,
          {
            id,
            wrapperClassName: 'datepicker--fluid',
            placeholderText: intl.formatMessage({ id: 'datetime.placeholder' }),
            selected: value,
            onChange: handleChangeDate(id),
            dateFormat: 'Pp',
            timeInputLabel: intl.formatMessage({ id: 'time' }),
            showTimeInput: true,
            locale: getLocale(),
          },
        ];
      default:
        return [];
    }
  }

  const [Component, componentProps] = getComponentByMetadata(metadata, value);

  if (Component != null) {
    return (
      <Field label={metadata.name} fieldId={metadata.id}>
        <Component {...componentProps} />
      </Field>
    );
  }
  return null;
}

DynamicField.propTypes = {
  metadata: PropTypes.object.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
  onChange: PropTypes.func,
};

export default React.memo(DynamicField);
