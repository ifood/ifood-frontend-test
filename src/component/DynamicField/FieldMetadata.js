import { formatDate, getLanguageName } from '../../core/l10n/L10n';

const SUPORTED_TYPE = {
  SELECT: Symbol('select'),
  DATE_TIME: Symbol('date_time'),
  NUMBER: Symbol('number'),
};

/**
 * TODO: This converter is not as dynamic as it should be
 */
const SPECIFIC_FIELD_CONVERTER = {
  locale: ({ name }) => getLanguageName(name),
};

const selectValuesMapper = ({ id, values }) => {
  return values.map((item) => {
    const formatter = SPECIFIC_FIELD_CONVERTER[id];
    let text = item.name;
    if (formatter != null) {
      const formatedText = formatter(item);
      if (formatedText != null) {
        text = formatedText;
      }
    }
    return {
      key: item.value,
      text,
      value: item.value,
    };
  });
};

const FIELD_FORMATTER = {
  [SUPORTED_TYPE.SELECT]: ({ fieldMetadata, value }) => {
    return selectValuesMapper(fieldMetadata).find(
      (item) => item?.value === value
    ).text;
  },
  [SUPORTED_TYPE.DATE_TIME]: ({ value }) =>
    value != null ? formatDate(value, 'Pp') : null,
};

const FIELD_SERIALIZER = {
  [SUPORTED_TYPE.DATE_TIME]: ({ value }) =>
    value != null ? value.toISOString() : null,
};

function getComponentType(metadata) {
  const { values, validation } = metadata;

  if (values != null) {
    return SUPORTED_TYPE.SELECT;
  }
  if (validation?.primitiveType === 'INTEGER') {
    return SUPORTED_TYPE.NUMBER;
  }
  if (validation?.entityType === 'DATE_TIME') {
    return SUPORTED_TYPE.DATE_TIME;
  }
  return null;
}

function getFieldFormatter(fieldMetadata) {
  const type = getComponentType(fieldMetadata);
  return FIELD_FORMATTER[type];
}

function getFieldSerializer(fieldMetadata) {
  const type = getComponentType(fieldMetadata);
  return FIELD_SERIALIZER[type];
}

function parseObjectByMetatada(metadata, object) {
  if (metadata == null || object == null) {
    return null;
  }
  return Object.keys(object).reduce((json, field) => {
    const fieldMeatada = metadata.find((item) => item.id === field);
    const serializer = getFieldSerializer(fieldMeatada);
    let value = object[field];
    if (serializer != null) {
      value = serializer({ value, fieldMeatada });
    }
    return {
      ...json,
      [field]: value,
    };
  }, {});
}

export {
  SUPORTED_TYPE,
  getFieldFormatter,
  getComponentType,
  parseObjectByMetatada,
  selectValuesMapper,
};
