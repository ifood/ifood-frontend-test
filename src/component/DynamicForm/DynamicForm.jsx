import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import DynamicField from '../DynamicField';

class DynamicForm extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      innerData: props.data,
    };
  }

  handleChange = ({ field, value }) => {
    this.setState((state) => {
      const innerData = { ...state.innerData };
      if (value == null || value === '') {
        delete innerData[field];
      } else {
        innerData[field] = value;
      }

      return {
        innerData,
      };
    });
  };

  handleSubmit = () => {
    const { onSubmit } = this.props;
    const { innerData } = this.state;
    if (onSubmit != null) {
      onSubmit(innerData);
    }
  };

  render() {
    const { id, metadata } = this.props;
    const { innerData } = this.state;
    if (metadata == null) {
      return null;
    }
    return (
      <Form id={id} onSubmit={this.handleSubmit}>
        {metadata.map((item) => (
          <DynamicField
            key={item.id}
            metadata={item}
            value={innerData?.[item.id]}
            onChange={this.handleChange}
          />
        ))}
      </Form>
    );
  }
}

DynamicForm.propTypes = {
  id: PropTypes.string,
  metadata: PropTypes.array,
  data: PropTypes.object,
  onSubmit: PropTypes.func,
};

export default DynamicForm;
