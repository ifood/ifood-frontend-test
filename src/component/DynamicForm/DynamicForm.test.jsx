import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import DynamicForm from './DynamicForm';
import DynamicField from '../DynamicField';

const metadata = [
  {
    id: 'limit',
    name: 'Quantidade',
    validation: {
      primitiveType: 'INTEGER',
      min: 1,
      max: 50,
    },
  },
];

const data = {
  limit: '10',
};

describe('<DynamicForm />', () => {
  it('snapshot', () => {
    const dynamicForm = mount(<DynamicForm metadata={metadata} />);
    jestExpect(toJson(dynamicForm)).toMatchSnapshot();
  });

  it('should return null if metadata is not provided', () => {
    const dynamicForm = shallow(<DynamicForm />);
    expect(dynamicForm.type()).to.be.null;
  });

  it('should use prop data as initial value for innerData', () => {
    const dynamicForm = shallow(<DynamicForm data={data} />);
    expect(dynamicForm.state().innerData).to.be.equal(data);
  });

  describe('when form is submited', () => {
    const onSubmit = jest.fn();
    const dynamicForm = mount(
      <DynamicForm data={data} metadata={metadata} onSubmit={onSubmit} />
    );

    it('should call onSubmit', () => {
      dynamicForm.find('form').simulate('submit');
      expect(onSubmit.mock.calls).to.have.lengthOf(1);
    });

    it('should call onSubmit with innerData as param', () => {
      dynamicForm.find('form').simulate('submit');
      expect(onSubmit.mock.calls[0][0]).to.be.equal(
        dynamicForm.state().innerData
      );
    });

    it('should not throw an error if onSubmit was not provided', () => {
      dynamicForm.setProps({ onSubmit: null });
      function submit() {
        dynamicForm.find('form').simulate('submit');
      }
      expect(submit).to.not.throw();
    });
  });

  describe('when a DynamicField is changed', () => {
    const dynamicForm = shallow(<DynamicForm metadata={metadata} />);
    const dynamicField = dynamicForm.find(DynamicField);
    const props = dynamicField.props();
    const field = props.metadata.id;
    const value = 'some value';

    it('should change internal state to the new value', () => {
      props.onChange({ field, value });
      expect(dynamicForm.state().innerData).to.deep.equal({
        [field]: value,
      });
    });

    it('should remove from internal state if new value is null', () => {
      props.onChange({ field });
      expect(dynamicForm.state().innerData).to.deep.equal({});
    });

    it('should remove from internal state if new value is an empty string', () => {
      props.onChange({ field, value: '' });
      expect(dynamicForm.state().innerData).to.deep.equal({});
    });
  });
});
