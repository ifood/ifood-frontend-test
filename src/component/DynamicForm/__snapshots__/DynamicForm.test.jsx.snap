// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicForm /> snapshot 1`] = `
<DynamicForm
  metadata={
    Array [
      Object {
        "id": "limit",
        "name": "Quantidade",
        "validation": Object {
          "max": 50,
          "min": 1,
          "primitiveType": "INTEGER",
        },
      },
    ]
  }
>
  <Form
    as="form"
    onSubmit={[Function]}
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <Memo(DynamicField)
        key="limit"
        metadata={
          Object {
            "id": "limit",
            "name": "Quantidade",
            "validation": Object {
              "max": 50,
              "min": 1,
              "primitiveType": "INTEGER",
            },
          }
        }
        onChange={[Function]}
      >
        <Field
          fieldId="limit"
          label="Quantidade"
        >
          <FormField>
            <div
              className="field"
            >
              <label
                htmlFor="limit"
              >
                Quantidade
              </label>
              <input
                id="limit"
                max={50}
                min={1}
                onChange={[Function]}
                placeholder="Quantidade"
                type="number"
                value=""
              />
            </div>
          </FormField>
        </Field>
      </Memo(DynamicField)>
    </form>
  </Form>
</DynamicForm>
`;
